### PR TITLE
[SPARK-29037][CORE][SQL] Fix the issue that spark gives duplicate result and support concurrent file source write operations write to different partitions in the same table.

### PR DESCRIPTION
**Pull request by turboFei**
## What changes were proposed in this pull request?
For InsertIntoHadoopFsRelation operations.

Case A:
Application appA insert overwrite table table_a with static partition overwrite.
But it was killed when committing tasks, because one task is hang.
And parts of its committed tasks output is kept under /path/table_a/_temporary/0/.

Then we rerun appA. It will reuse the staging dir /path/table_a/_temporary/0/.
It executes successfully.
But it also commit the data reminded by killed application to destination dir.

Case B:

Application appA insert overwrite table table_a.

Application appB insert overwrite table table_a, too.

They execute concurrently, and they may all use /path/table_a/_temporary/0/ as workPath.

And their result may be corruptted.

In this PR, we set a staging output path for InsertIntoHadoopFsRelation operation.

The output path is a multi level path and is composed of specified partition key-value pairs formatted .spark-staging-${depth}/p1=v1/p2=v2/.../pn=vn.

We can detect the conflict by checking the existence of relative staging output path.

For example:

table_a(c1 string, a string, b string,...) partitioned by (a, b , ...)

When writing to partition a=1, we need to check the existence of

.spark-staging-0
.spark-staging-1/a=1
.spark-staging-2/a=1
...
.spark-staging-numPartitions/a=1
When we found relative staging output path is existed, an exception would be thrown.

Besides, we propose an operation liveness detection mechanism to automatically detect whether an operation is running based on hdfs lease mechanism.

Thanks @advancedxy for providing the prototype of solution and valuable suggestions.

## Why are the changes needed?
Data may be corrupted without this PR.

## Does this PR introduce any user-facing change?
User can see the exists of staging output path and may need clean up these path manually, when this path is belong to a killed application and has not been cleaned up gracefully.

## How was this patch tested?
Added unit test.

### CODE REVIEW

**Comment by turboFei** [comment]
just cc @advancedxy

**Comment by turboFei** [comment]
We propose an operation liveness detection mechanism to automatically detect whether an operation is running based on hdfs lease mechanism, and automatically clean up the conflicted staging partition path, which has been cleaned up gracefully.

**Comment by turboFei** [comment]
Besides, In this PR, we just set an output dir for FileOutputCommitter and does not invoke any methods of FileOutputCommitter.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala [hunk 1]
```diff
@@ -106,103 +120,143 @@ case class InsertIntoHadoopFsRelationCommand(
         fs, catalogTable.get, qualifiedOutputPath, matchingPartitions)
     }
 
+    val parameters = CaseInsensitiveMap(options)
+
+    val partitionOverwriteMode = parameters.get("partitionOverwriteMode")
+      // scalastyle:off caselocale
+      .map(mode => PartitionOverwriteMode.withName(mode.toUpperCase))
```
Comment (advancedxy) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala]: Please don't turn off caselocale style check, just call `mode.toUpperCase(Locale.ROOT)`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala [hunk 2]
```diff
@@ -269,4 +323,214 @@ case class InsertIntoHadoopFsRelationCommand(
...
     }.toMap
   }
+
+  /**
+   * Check current committer whether supports several InsertIntoHadoopFsRelation operations write
+   * to different partitions in a same table concurrently. If supports, then detect the conflict
+   * whether there are several operations write to same partition in the same table or write to
+   * a non-partitioned table.
+   */
+  private def detectConflict(
```
Comment (advancedxy) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala]: Kind of feeling that this is more suitable inside of FileCommitProtocol. Let's wait for others opinion

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala [hunk 3]
```diff
@@ -269,4 +323,214 @@ case class InsertIntoHadoopFsRelationCommand(
...
+          }
+        } match {
+          case Success(_) => true
+          case _ => false
+        }
+
+        if (pickLock) {
+          true
+        } else {
+          Thread.sleep(HdfsConstants.LEASE_SOFTLIMIT_PERIOD)
```
Comment (advancedxy) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala]: This is quite some time(IIRC, 60secs), should we add an option to turn off this wait?

**Comment by advancedxy** [comment]
Gently ping @cloud-fan .

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala (hunk 1 above)
Comment (turboFei) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala]: I will rebase this part later.

**Comment by turboFei** [comment]
## About HDFS lease and soft limit period
HDFS is an distributed file system, it has a lease mechanism to maintain lots of DFSClients to access files.

There is an constant filed LEASE_SOFTLIMIT_PERIOD, whose value is 60 seconds.



DFSClient get a lease of an file and it would be added into the dfsclient list of LeaseRenewer
LeaseRenewer would refresh the leases with a period LEASE_SOFTLIMIT_PERIOD/2
When DFSClient closed or aborted, it would be removed from the dfsClient list of LeaseRenewer and has no chance to refresh its leases


If a dfsclient want to write a file, this file must be in immutable status.

When a dfsClient begin to write a file, this file moved to being written status.



If a dfsCilent begin to write a file, but its relative lease has not been refreshed for a LEASE_SOFTLIMIT_PERIOD,  nameNode would move this file that is being written to be immutable.

It means that,  another dfsClient can begin to write this file, such as append operation.



## Operation liveness detection mechanism
So, we propose an operation liveness detection mechanism based it:

For each operation, we create a LOCK file under its staging path and keep this LOCK file being written status.
After this operation completed, deleing this staging path and the lock file under it would be released. 
Even this operation aborted and could not cleanup this staging path gracefully, the LOCK file would be moved to immutable status after 60 seconds(LEASE_SOFT_LIMIT_PERIOD).
When detecting some conflicted staging path, we try to get relative LOCK under staging path, if successful, delete this staging path automatically, otherwise, the operation owns this staging path is still running and we need wait it to complete.
When we try to get relative LOCK under staging path, if this LOCK file is not existed, we create it and get the LOCK; otherwise, we try to append this LOCK file, if append successfully, we get the LOCK successful, otherwise, it is owned by another DFSClient(operation).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala (hunk 3 above)
Comment (turboFei) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala]: Here we just double check after a soft limit period when the first check is failed.

In fact, we can set an timeout and try to pick lock with a period, such as 1000 ms.

**Pull request closed without merging**